axi_intercon_gen.py

```diff
@@ -0,0 +1,552 @@
+#!/usr/bin/env python
+import math
+import sys
+from collections import OrderedDict, defaultdict
+import yaml
+
+from verilogwriter import Signal, Wire, Instance, ModulePort, Port, VerilogWriter
+
+if sys.version[0] == '2':
+
+    math.log2 = lambda x : math.log(x, 2)
+
+class Widths:
+    addr = 0
+    user = 0
+    data = 0
+    max_id = 0
+
+def axi_signals2(w, id_width):
+    signals = [
+        ("aw", "id"    , False, id_width),
+        ("aw", "addr"  , False, w.addr  ),
+        ("aw", "len"   , False, 8  ),
+        ("aw", "size"  , False, 3  ),
+        ("aw", "burst" , False, 2 ),
+        ("aw", "lock"  , False, 0 ),
+        ("aw", "cache" , False, 4 ),
+        ("aw", "prot"  , False, 3 ),
+        ("aw", "region", False, 4),
+        ("aw", "user"  , False, w.user),
+        ("aw", "qos"   , False, 4),
+        ("aw", "valid" , False, 0),
+        ("aw", "ready" , True , 0),
+
+        ("ar", "id"    , False, id_width),
+        ("ar", "addr"  , False, w.addr),
+        ("ar", "len"   , False, 8),
+        ("ar", "size"  , False, 3),
+        ("ar", "burst" , False, 2),
+        ("ar", "lock"  , False, 0),
+        ("ar", "cache" , False, 4),
+        ("ar", "prot"  , False, 3),
+        ("ar", "region", False, 4),
+        ("ar", "user"  , False, w.user),
+        ("ar", "qos"   , False, 4),
+        ("ar", "valid" , False, 0),
+        ("ar", "ready" , True , 0),
+
+        ("w", "data" , False, w.data),
+        ("w", "strb" , False, w.data//8),
+        ("w", "last" , False, 0),
+        ("w", "user"  , False, w.user),
+        ("w", "valid", False, 0),
+        ("w", "ready", True , 0),
+
+        ("b", "id"   , True , id_width),
+        ("b", "resp" , True , 2),
+        ("b", "valid", True , 0),
+        ("b", "user"  , True, w.user),
+        ("b", "ready", False, 0),
+
+        ("r", "id"   , True , id_width),
+        ("r", "data" , True , w.data),
+        ("r", "resp" , True , 2),
+        ("r", "last" , True , 0),
+        ("r", "user"  , True, w.user),
+        ("r", "valid", True , 0),
+        ("r", "ready", False, 0),
+    ]
+    return signals
+
+def axi_signals(w, id_width, drop_prefix=False):
+    signals = [
+        ("awid"    , False, id_width),
+        ("awaddr"  , False, w.addr  ),
+        ("awlen"   , False, 8  ),
+        ("awsize"  , False, 3  ),
+        ("awburst" , False, 2 ),
+        ("awlock"  , False, 0 ),
+        ("awcache" , False, 4 ),
+        ("awprot"  , False, 3 ),
+        ("awregion", False, 4),
```

Comment:
We have an additional `awatop` port for AXI5 ATOPs, does that need to be added to this list?

Reply:
I ignored awatop, but you're right. It should be added. Since it's probably not commonly used, I would prefer to have it optional. I can add a config switch to enable it